File: firmware/WakeLink/packet.cpp
#include "packet.h"
#include "platform.h"

extern DeviceConfig cfg;
extern String DEVICE_ID;

/**
 * @brief Generate short unique request identifier.
 *
 * Creates simple 8-character ID from limited character set.
 * Used to bind response to request.
 *
 * @return Request ID string.
 */
String PacketManager::generateRequestId() {
    const char* chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    String id = "";
    for (int i = 0; i < 8; ++i) {
        id += chars[random(0, 36)];
    }
    return id;
}

/**
 * @brief Create signed command packet.
 *
 * Assembles internal JSON with command, data, request_id, timestamp.
 * Encrypts internal JSON and forms outer packet via createOuterPacket.
 *
 * @param command Command name.
 * @param data Command data object.
 * @return Serialized signed outer packet.
 */
String PacketManager::createCommandPacket(const String& command, const JsonObject& data) {
    JsonDocument innerDoc;
    innerDoc["command"] = command;
    innerDoc["data"] = data;
    innerDoc["request_id"] = generateRequestId();
    innerDoc["timestamp"] = millis();

    String encryptedPayload = encryptJson(innerDoc);
    return createOuterPacket(encryptedPayload);
}

/**
 * @brief Create outer JSON packet.
 *
 * Forms JSON with device_id, payload, signature, and version fields.
 * Signature is computed via crypto.calculateHMAC for payload.
 *
 * @param encryptedPayload Hex-encoded encrypted payload.
 * @return Serialized outer JSON packet.
 */
String PacketManager::createOuterPacket(const String& encryptedPayload) {
    JsonDocument doc;
    doc["device_id"] = DEVICE_ID;
    doc["payload"] = encryptedPayload;
    doc["signature"] = crypto.calculateHMAC(encryptedPayload);
    doc["version"] = "1.0";

    String out;
    serializeJson(doc, out);
    return out;
}

/**
 * @brief Parse outer JSON packet.
 *
 * Deserializes outer JSON, validates required fields and version.
 * Verifies HMAC signature using crypto.verifyHMAC.
 * Returns JsonDocument with status and encrypted_payload or error.
 *
 * @param packet Raw outer JSON packet.
 * @return JsonDocument with parsing result.
 */
JsonDocument PacketManager::parseOuterPacket(const String& packet) {
    JsonDocument doc, result;

    if (deserializeJson(doc, packet)) {
        result["status"] = "error";
        result["error"] = "JSON_PARSE";
        return result;
    }

    String payload = doc["payload"] | "";
    String sig = doc["signature"] | "";
    String version = doc["version"] | "";

    if (version != "1.0" || payload.isEmpty() || sig.isEmpty()) {
        result["status"] = "error";
        result["error"] = "BAD_PACKET";
        return result;
    }

    if (!crypto.verifyHMAC(payload, sig)) {
        Serial.printf("[SIGN] Expected: %s\n", crypto.calculateHMAC(payload).c_str());
        Serial.printf("[SIGN] Received: %s\n", sig.c_str());
        result["status"] = "error";
        result["error"] = "INVALID_SIGNATURE";
        return result;
    }

    Serial.println("[SIGN] Signature OK");

    result["status"] = "success";
    result["encrypted_payload"] = payload;
    return result;
}

/**
 * @brief Process incoming encrypted packet.
 *
 * Parses outer packet, decrypts internal JSON via crypto.processSecurePacket.
 * Validates internal JSON structure (presence of command field).
 * Returns JsonDocument with status and data/error.
 *
 * @param packetData Raw incoming packet.
 * @return JsonDocument with processing result.
 */
JsonDocument PacketManager::processIncomingPacket(const String& packetData) {
    JsonDocument result;
    
    JsonDocument outerResult = parseOuterPacket(packetData);
    if (outerResult["status"] != "success") {
        return outerResult;
    }
    
    String encryptedPayload = outerResult["encrypted_payload"];
    
    String decrypted = crypto.processSecurePacket(encryptedPayload);
    if (decrypted.startsWith("ERROR:")) {
        result["status"] = "error";
        result["error"] = decrypted;
        return result;
    }
    
    DeserializationError error = deserializeJson(result, decrypted);
    if (error) {
        result["status"] = "error";
        result["error"] = "INVALID_JSON";
        result["raw_error"] = error.c_str();
        return result;
    }
    
    if (result["command"].isNull()) {
        result["status"] = "error";
        result["error"] = "NO_COMMAND";
        return result;
    }
    
    JsonVariant dataVar = result["data"];
    if (dataVar.isNull() || !dataVar.is<JsonObject>()) {
        result["data"].to<JsonObject>();
    }
    
    result["status"] = "success";
    return result;
}

/**
 * @brief Create encrypted response packet.
 *
 * Encrypts JsonDocument result and forms outer signed packet.
 *
 * @param resultData Result data to send.
 * @return Serialized signed outer packet.
 */
String PacketManager::createResponsePacket(const JsonDocument& resultData) {
    String encryptedPayload = encryptJson(resultData);
    return createOuterPacket(encryptedPayload);
}

/**
 * @brief Encrypt JSON document.
 *
 * Serializes json to string and calls crypto.createSecureResponse to get hex packet.
 *
 * @param json JSON document to encrypt.
 * @return Hex-encoded encrypted payload.
 */
String PacketManager::encryptJson(const JsonDocument& json) {
    String plaintext;
    serializeJson(json, plaintext);
    return crypto.createSecureResponse(plaintext);
}

File: client/core/protocol/packet.py
"""
WakeLink Protocol v1.0 Packet Manager.

Handles creation and processing of signed, encrypted packets for
communication with WakeLink devices. Compatible with firmware packet.cpp.

Packet format:
- Outer JSON: {"device_id", "payload", "signature", "version"}
- Payload: hex string = [uint16_be length] + [ciphertext] + [16-byte nonce]
- Signature: HMAC-SHA256 of payload hex string only

The server acts as a transparent relay and never decrypts the payload.
"""

import json
import time
import uuid
from typing import Any, Dict, Optional

from ..crypto import Crypto


class PacketManager:
    """Manages packet creation and processing for WakeLink protocol v1.0.
    
    Attributes:
        crypto: Crypto instance for encryption/signing.
        device_id: Device identifier for packet headers.
    """
    
    PROTOCOL_VERSION = "1.0"
    
    def __init__(self, token: str, device_id: str):
        """Initialize packet manager.
        
        Args:
            token: Device token (min 32 chars) for key derivation.
            device_id: Device identifier for packet headers.
        """
        self.crypto = Crypto(token)
        self.device_id = device_id
    
    def create_command_packet(self, command: str, data: Optional[Dict[str, Any]] = None) -> str:
        """Create a signed, encrypted command packet.
        
        Args:
            command: Command name (e.g., "ping", "wake", "info").
            data: Optional command parameters.
            
        Returns:
            JSON string of outer packet ready for transmission.
        """
        # Build inner packet
        inner = {
            "command": command,
            "data": data or {},
            "request_id": str(uuid.uuid4())[:8],
            "timestamp": int(time.time())
        }
        
        # Encrypt inner packet
        inner_json = json.dumps(inner, separators=(",", ":"))
        payload_hex = self.crypto.create_secure_response(inner_json)
        
        # Calculate HMAC on payload only
        signature = self.crypto.calculate_hmac(payload_hex)
        
        # Build outer packet
        outer = {
            "device_id": self.device_id,
            "payload": payload_hex,
            "signature": signature,
            "version": self.PROTOCOL_VERSION
        }
        
        return json.dumps(outer, separators=(",", ":"))
    
    def process_incoming_packet(self, packet_json: str) -> Dict[str, Any]:
        """Process an incoming signed, encrypted packet.
        
        Args:
            packet_json: JSON string of outer packet.
            
        Returns:
            Dict with status, decrypted data, and counter from ESP.
        """
        try:
            # Parse outer packet
            outer = json.loads(packet_json)
        except json.JSONDecodeError as e:
            return {"status": "error", "error": f"JSON_PARSE_ERROR: {e}"}
        
        # Validate required fields
        required = ["device_id", "payload", "signature"]
        for field in required:
            if field not in outer:
                return {"status": "error", "error": f"MISSING_FIELD: {field}"}
        
        payload_hex = outer["payload"]
        signature = outer["signature"]
        
        # Verify HMAC signature (on payload only)
        if not self.crypto.verify_hmac(payload_hex, signature):
            return {"status": "error", "error": "INVALID_SIGNATURE"}
        
        # Decrypt payload
        decrypted = self.crypto.process_secure_packet(payload_hex)
        
        if decrypted.startswith("ERROR:"):
            return {"status": "error", "error": decrypted}
        
        # Parse inner packet
        try:
            inner = json.loads(decrypted)
        except json.JSONDecodeError as e:
            return {"status": "error", "error": f"INNER_JSON_ERROR: {e}"}
        
        # Build successful response with counter from outer packet
        result = {"status": "success"}
        
        # Add counter from outer packet (synced from ESP)
        if "counter" in outer:
            result["counter"] = outer["counter"]
        
        result.update(inner)
        
        return result
    
    def create_response_packet(self, response_data: Dict[str, Any]) -> str:
        """Create a signed, encrypted response packet.
        
        Used by the device to send responses back to the client.
        
        Args:
            response_data: Response data to encrypt and sign.
            
        Returns:
            JSON string of outer packet.
        """
        # Add timestamp if not present
        if "timestamp" not in response_data:
            response_data["timestamp"] = int(time.time())
        
        # Encrypt response
        inner_json = json.dumps(response_data, separators=(",", ":"))
        payload_hex = self.crypto.create_secure_response(inner_json)
        
        # Calculate HMAC on payload only
        signature = self.crypto.calculate_hmac(payload_hex)
        
        # Build outer packet
        outer = {
            "device_id": self.device_id,
            "payload": payload_hex,
            "signature": signature,
            "version": self.PROTOCOL_VERSION
        }
        
        return json.dumps(outer, separators=(",", ":"))
